Dockerfile

```diff
@@ -1,44 +1,200 @@
+    PATH="${PATH}:/home/${USER}/.local/bin:${GEOIPS_DEPENDENCIES_DIR}/bin"
 
-WORKDIR $GEOIPS_PACKAGES_DIR
+# Create a non-root user
+RUN groupadd -g "${GROUP_ID}" "${USER}" \
+    && useradd -l -m -u "${USER_ID}" -g "${GROUP_ID}" "${USER}"
 
-COPY ./environments/requirements.txt ${GEOIPS_PACKAGES_DIR}/requirements.txt
 
-RUN python3 -m pip install -r requirements.txt
+# Create directories; set ownership and permissions to all if UNSAFE_PERMS != "False"
+RUN mkdir -p "${GEOIPS_OUTDIRS}" \
+    && mkdir -p "${GEOIPS_DEPENDENCIES_DIR}" \
+    && mkdir -p "${GEOIPS_TESTDATA_DIR}" \
+    && mkdir -p "${GEOIPS_PACKAGES_DIR}/geoips" \
+    && chown -R "${USER}:${GROUP_ID}" \
+           "${GEOIPS_OUTDIRS}" \
+           "${GEOIPS_DEPENDENCIES_DIR}" \
+           "${GEOIPS_TESTDATA_DIR}" \
+           "${GEOIPS_PACKAGES_DIR}" \
+    && if [ "${UNSAFE_PERMS}" != "False" ]; then \
+         chmod -R a+rw \
+           "${GEOIPS_OUTDIRS}" \
+           "${GEOIPS_DEPENDENCIES_DIR}" \
+           "${GEOIPS_TESTDATA_DIR}" \
+           "${GEOIPS_PACKAGES_DIR}"; \
+       fi
 
+USER ${USER}
+
+# Set working directory
 WORKDIR ${GEOIPS_PACKAGES_DIR}/geoips
-COPY . .
 
-RUN pip install -e ".[test, doc, lint]" \
+# Copy only requirements first to leverage Docker layer caching
+COPY --chown=${USER}:${GROUP_ID} ${GEOIPS_BASE_CLONE_DIR}/environments/requirements.txt ./requirements.txt
+RUN python -m pip install --no-cache-dir -r requirements.txt
+
+# Now copy all of GeoIPS
+COPY --chown=${USER}:${GROUP_ID} ${GEOIPS_BASE_CLONE_DIR} ./
+
+# Configure Git (avoid "detected dubious ownership" warnings when mounted)
+RUN git config --global --add safe.directory '*'
+
+# Install GeoIPS in editable mode
+RUN python -m pip install --no-cache-dir -e "." \
     && create_plugin_registries
+
+###############################################################################
+#                          FULL BUILD STAGE
+###############################################################################
+FROM build AS full_build
+
+# Install all plugins
+RUN python -m pip install --no-cache-dir -e "$GEOIPS_PACKAGES_DIR/geoips/" \
+    && bash $GEOIPS_PACKAGES_DIR/geoips/tests/integration_tests/base_install.sh \
+    && bash $GEOIPS_PACKAGES_DIR/geoips/tests/integration_tests/full_install.sh \
+    && create_plugin_registries
+
+###############################################################################
+#                          SYSTEM BUILD STAGE
+###############################################################################
+FROM build AS system_build
+
+# Install all plugins
+RUN python -m pip install --no-cache-dir -e "$GEOIPS_PACKAGES_DIR/geoips/" \
+    && bash $GEOIPS_PACKAGES_DIR/geoips/tests/integration_tests/base_install.sh \
+    && bash $GEOIPS_PACKAGES_DIR/geoips/tests/integration_tests/full_install.sh \
+    && bash $GEOIPS_PACKAGES_DIR/geoips/tests/integration_tests/system_install.sh \
+    && create_plugin_registries
+
```

Comment:
Should this stage build off of the `full_build` stage and just call `bash $GEOIPS_PACKAGES_DIR/geoips/tests/integration_tests/system_install.sh; create_plugin_registries`

Reply:
Sure, that'd be fine with me - this results in a smaller image by running them all together but not by much

Reply:
IMO this is more of a problem with our install then with the dockerfile